packages/aws-cdk/lib/api/cxapp/cloud-executable.ts

```diff
@@ -89,54 +90,75 @@ export class CloudExecutable {
+        await this.addMetadataResource(assembly);
       }
 
       return new CloudAssembly(assembly);
+    }
+  }
+
+  /**
+   * Modify the templates in the assembly in-place to add metadata resource declarations
+   */
+  private async addMetadataResource(rootAssembly: cxapi.CloudAssembly) {
+    if (!rootAssembly.runtime) { return; }
+
+    const modules = formatModules(rootAssembly.runtime);
+    await processAssembly(rootAssembly);
 
-      function formatModules(runtime: cxapi.RuntimeInfo): string {
-        const modules = new Array<string>();
+    async function processAssembly(assembly: cxapi.CloudAssembly) {
+      for (const stack of assembly.stacks) {
+        await processStack(stack);
+      }
+      for (const nested of assembly.nestedAssemblies) {
+        await processAssembly(nested.nestedAssembly);
+      }
+    }
 
-        // inject toolkit version to list of modules
-        // eslint-disable-next-line @typescript-eslint/no-require-imports
-        const toolkitVersion = require('../../../package.json').version;
-        modules.push(`aws-cdk=${toolkitVersion}`);
+    async function processStack(stack: cxapi.CloudFormationStackArtifact) {
+      const resourcePresent = stack.environment.region === cxapi.UNKNOWN_REGION
+        || RegionInfo.get(stack.environment.region).cdkMetadataResourceAvailable;
+      if (!resourcePresent) { return; }
 
-        for (const key of Object.keys(runtime.libraries).sort()) {
-          modules.push(`${key}=${runtime.libraries[key]}`);
+      if (!stack.template.Resources) {
+        stack.template.Resources = {};
+      }
+      if (stack.template.Resources.CDKMetadata) {
+        warning(`The stack ${stack.id} already includes a CDKMetadata resource`);
+        return;
+      }
+
+      stack.template.Resources.CDKMetadata = {
+        Type: 'AWS::CDK::Metadata',
+        Properties: {
+          Modules: modules,
+        },
+      };
+
+      if (stack.environment.region === cxapi.UNKNOWN_REGION) {
+        stack.template.Conditions = stack.template.Conditions || {};
+        const condName = 'CDKMetadataAvailable';
+        if (!stack.template.Conditions[condName]) {
+          stack.template.Conditions[condName] = _makeCdkMetadataAvailableCondition();
+          stack.template.Resources.CDKMetadata.Condition = condName;
+        } else {
+          warning(`The stack ${stack.id} already includes a ${condName} condition`);
         }
-        return modules.join(',');
       }
+
+      // The template has changed in-memory, but the file on disk remains unchanged so far.
+      // The CLI *might* later on deploy the in-memory version (if it's <50kB) or use the
+      // on-disk version (if it's >50kB).
+      //
+      // Be sure to flush the changes we just made back to disk. The on-disk format is always
+      // JSON.
+      await fs.writeFile(stack.templateFullPath, JSON.stringify(stack.template, undefined, 2), { encoding: 'utf-8' });
```

Comment:
Mutate the synthesized template. Ewww...

I suppose the root problem is that the CLI shouldn't be injecting the metadata. Is it safe to assume this will not happen when the correct fix in put in place?